cdap-ui/app/directives/dag-plus/my-dag-ctrl.js

```diff
@@ -454,29 +313,181 @@ angular.module(PKG.name + '.commons')
       };
     }
 
-    function formatConnections() {
-      var connections = [];
-      angular.forEach(vm.instance.getConnections(), function (conn) {
-        connections.push({
-          from: conn.sourceId,
-          to: conn.targetId
-        });
+    function addConnection(newConnObj) {
+      $scope.connections.push({
+        from: newConnObj.sourceId,
+        to: newConnObj.targetId
+      });
+      DAGPlusPlusNodesActionsFactory.setConnections($scope.connections);
+    }
+
+    function removeConnection(detachedConnObj, updateStore = true) {
+      var detachedConnection = _.find($scope.connections, function (conn) {
+        return conn.from === detachedConnObj.sourceId && conn.to === detachedConnObj.targetId;
+      });
+      if (detachedConnection) {
+        $scope.connections.splice($scope.connections.indexOf(detachedConnection), 1);
+      }
+      if (updateStore) {
+        DAGPlusPlusNodesActionsFactory.setConnections($scope.connections);
+      }
+    }
+
+    function moveConnection(moveInfo) {
+      let oldConnection = {
+        sourceId: moveInfo.originalSourceId,
+        targetId: moveInfo.originalTargetId
+      };
+      // don't need to call addConnection for the new connection, since that will be done
+      // automatically as part of the 'connection' event
+      removeConnection(oldConnection, false);
+    }
+
+    vm.removeSelectedConnections = function() {
+      if (selectedConnections.length === 0 || vm.isDisabled) { return; }
+
+      angular.forEach(selectedConnections, function (selectedConnectionObj) {
+        vm.instance.unbind('connectionDetached');
+        removeContextMenuEventListener(selectedConnectionObj);
+        vm.instance.detach(selectedConnectionObj);
+        removeConnection(selectedConnectionObj, false);
+        vm.instance.bind('connectionDetached', removeConnection);
+      });
+      selectedConnections = [];
+      DAGPlusPlusNodesActionsFactory.setConnections($scope.connections);
+    };
+
+    function selectEndpointOrConnection(seletectedObj) {
+      // is endpoint
+      if (seletectedObj.endpoint) {
+        if (seletectedObj.connections && seletectedObj.connections.length > 0) {
+          let connections = vm.instance.getConnections({
+            source: seletectedObj.connections[0].sourceId
+          });
+
+          // This is to toggle all connections coming from an endpoint.
+          // If zero, one or more (but not all) of the connections are already selected,
+          // then just select the remaining ones. Else if they're all selected,
+          // then unselect them.
+
+          if (_.difference(connections, selectedConnections).length !== 0) {
```

Comment:
Wouldn't lodash's `difference` function already return the difference between selected connections and all the connections? Can we use that to iterate the rest of the connections to be selected? Could iterate over the difference rather than iterating over connections and checking if its already part of selectedConnections